enhancements/installer/ipi-publishing-status.md

```diff
@@ -0,0 +1,206 @@
+
+Exposing the cluster's publishing status empowers other components to bootstrap
+themselves as public or private according to the user request at install time.
+
+## Motivation
+
+A component should be able to know how to bootstrap itself according to user
+preference. A central place providing the cluster publishing status at install
+time makes more sense than the installer having to individually configure
+various components.
+
+### User Stories
+
+* As the author of the image registry operator, I want to easily discover the
+cluster publishing status, so that I can configure access the cloud storage objects
+accordingly.
+
+### Goals
+
+Expose the cluster publishing status used at install time in the Infrastructure
+object.
+
+### Non-Goals
+
+This enhancement does not cover support for updating the cluster publishing
+status on running clusters. This property is meant to reflect the `publish`
+parameter as it was set in the install-config.yaml.
+
+## Proposal
+
+The Infrastructure API is extended, adding the `Publish` field to the
+`InfrastructureStatus`.
+
+```go
+type InfrastructureStatus struct {
+	// ...
+
+	// publish indicates the publishing status of the cluster at install time.
+	// The default is 'External', which means the cluster will be publicly
+    // available on the internet.
+	// The 'Internal' mode indicates a private cluster, and operators should
+    // configure operands accordingly.
+    // +kubebuilder:default=External
+    // +kubebuilder:validation:Enum=Internal;External
+	Publish PublishMode `json:"publish"`
+}
+```
+
+To reflect the install-config publishing options, `PublishMode` will only
+support to valid options.
+
+```
+// PublishMode defines the publish mode of the cluster at install time.
+type PublishMode string
+
+const (
+	// "Internal" is for operators to configure their asserts as private
+    // when appropriate.
+	InternalPublishMode PublishMode = "Internal"
+
+	// "External" indicates that the cluster is publicly available on the internet.
+	ExternalPublishMode PublishMode = "External"
+)
+```
+
+### Workflow Description
+
+1. The cluster creator install a cluster via IPI, setting `publish: Internal` in installer-config.yaml
+2. The installer bootstraps the cluster, and sets `Publish: Internal` in the Infrastructure status
+3. Operators use this value to configure operands accordingly.
+
+### API Extensions
+
+### Implementation Details/Notes/Constraints [optional]
+
+What are the caveats to the implementation? What are some important details that
+didn't come across above. Go in to as much detail as necessary here. This might
+be a good place to talk about core concepts and how they relate.
+
+The `Publish` field in the infrastructure status will only reflect the initial
+status of the cluster. Individual components (like ingress) may be changed on
+day-2 to for example turn a public cluster private, and the `Publish` field in
+the Infrastructure status will remain untouched.
```

Comment:
What's involved in converting a cluster from public to private? Is there a documented solution for this?

Reply:
Day-2 conversion of a cluster from public to private involves manually configuring components individually. The process is described in [OCP docs[1]](https://docs.openshift.com/container-platform/4.13/post_installation_configuration/configuring-private-cluster.html).

[1] https://docs.openshift.com/container-platform/4.13/post_installation_configuration/configuring-private-cluster.html

Reply:
It may be helpful to include that link within the EP.

Typically a publish field like this would be the source of truth for whether a cluster is public or private. Have the cluster half public and half private sounds like it could be confusing to understand and debug not only for end users, but also for support. Is this half and half scenario something we want to continue to support through this EP, or perhaps we need to breakdown the publishing status so there is still a source of truth, but it's per component.

Eg if the spec had an ingressController.Publish field that the ingress operator could use to determine if it should be public or private, that might make things very clear and automate some of these manual steps away

Reply:
Yeah we might need to rename the field so that it's more clear that this is the initial publishing status (hey that might be a good name!) as to avoid confusion.

As for the half public and half private thing, it's more of a thing than I previously knew - ARO and Hypershift both support a combination of private and public along with the regular all public and all private installs. So this isn't only a day-2 thing. See https://issues.redhat.com/browse/CORS-2743.